lib/Driver/UnixToolChains.cpp

```diff
@@ -182,27 +182,28 @@ toolchains::GenericUnix::constructInvocation(const DynamicLinkJobAction &job,
 
   // Configure the toolchain.
   //
-  // By default use the system `clang` to perform the link.  We use `clang` for
-  // the driver here because we do not wish to select a particular C++ runtime.
-  // Furthermore, until C++ interop is enabled, we cannot have a dependency on
-  // C++ code from pure Swift code.  If linked libraries are C++ based, they
-  // should properly link C++.  In the case of static linking, the user can
-  // explicitly specify the C++ runtime to link against.  This is particularly
-  // important for platforms like android where as it is a Linux platform, the
-  // default C++ runtime is `libstdc++` which is unsupported on the target but
-  // as the builds are usually cross-compiled from Linux, libstdc++ is going to
-  // be present.  This results in linking the wrong version of libstdc++
-  // generating invalid binaries.  It is also possible to use different C++
-  // runtimes than the default C++ runtime for the platform (e.g. libc++ on
-  // Windows rather than msvcprt).  When C++ interop is enabled, we will need to
-  // surface this via a driver flag.  For now, opt for the simpler approach of
-  // just using `clang` and avoid a dependency on the C++ runtime.
-  const char *Clang = "clang";
+  // We don't use `clang++` unconditionally because we want to avoid pulling in
+  // a C++ standard library if it's not needed, in particular because the
+  // standard library that `clang++` selects by default may not be the one that
+  // is desired.
+  //
+  // TODO: In principle, it should be possible to use a different C++ standard
+  // library than the one configured by default by passing a `-stdlib` option
+  // to `-Xcc` and `-Xclang-linker`, e.g.
+  // `-Xcc -stdlib=libc++ -Xclang-linker -stdlib=libc++`.
+  // Once there is a driver flag for C++ interop, we will probably also want to
+  // add a driver flag for selecting the C++ standard library.
+  //
+  // In practice, using libc++ on Linux, for example, does not work because the
+  // SwiftGlibc module definition is incompatible with libc++'s header layout.
+  // We probably need to ensure that we use libc++'s own module map instead of 
+  // the SwiftGlibc module map.
```

Comment:
Do you mean I need to pass `-fmodules-local-submodule-visibility`?

I've tried that, but it doesn't work -- and I don't really see how it would make a difference, as the error I'm getting is a cyclic dependency between modules `std_config` (from libc++'s module map) and `SwiftGlibc`, and IIUC, submodule visibility shouldn't have an influence on this?

Anyway, I've deleted the whole comment referring to libc++ and SwiftGlibc, so this should no longer be an issue for this PR, but I'm still interested in this problem more generally. I'll post a gist later containing the test I'm running and the output I'm getting, and I'd be interested in your insights.

Reply:
Update: Here's a gist with a test that shows the problem, as well as the error messages produced by Swift:

https://gist.github.com/martinboehme/9faa7747ba24d29f6f0f6915bb3dcb7f

As the error messages note, there's a cyclic dependency between the modules `std_config` (from libc++) and `SwiftGlibc`.

This cyclic dependency comes about as follows:

*  `/usr/include/c++/v1/__config` (from libc++'s module `std_config`) includes `/usr/include/features.h` (from the module `SwiftGlibc`)
*  `/usr/include/inttypes.h` (from `SwiftGlibc`) includes `<stdint.h>` (which because of include paths resolves to `/usr/include/c++/v1/stdint.h`), which includes `/usr/include/c++/v1/__config` (from`std_config`)

Note that the cyclic dependency between the modules, not on the level of files, i.e. it's not the case that `__config` includes a `SwiftGlibc` header file that then (possibly indirectly) includes `__config` again.

It seems that a solution might be to move `features.h` to a different top-level module outside the `SwiftGlibc` module. `__config` doesn't include much else, so this should break the cycle, and `features.h` isn't a standard C header anyway -- IIUC it's really an implementation detail of glibc, so it would be cleaner in any case not to expose it from `SwiftGlibc`.

WDYT?

Reply:
That is a problem we actually need to solve.  The modules are already highly opaque, and this only makes it more complicated.  Actually, SwiftGlibc should not have any dependencies on `std`, just on `Glibc`.  What you are describing sounds like a failure of the `Glibc` module to fully modularlize the C library and having parts of the C library escape into the newly introduced C++ module.

Reply:
> That is a problem we actually need to solve.

Agree. @gribozavr and I are working on this and have some ideas, though they're still a work in progress -- but we expect to post a proposal on the forum soon.

> The modules are already highly opaque, and this only makes it more complicated. Actually, SwiftGlibc should not have any dependencies on `std`, just on `Glibc`. What you are describing sounds like a failure of the `Glibc` module to fully modularlize the C library and having parts of the C library escape into the newly introduced C++ module.

Unfortunately, it's more tricky than that. The problem is that, fundamentally, glibc and libc++ aren't layered with respect to each other because glibc inadvertently ends up including libc++'s wrapper headers. See the example I discuss above for `stdint.h.`.

The options are:
*  Modify glibc
    * This seems infeasible, as we have to be compatible with existing distributions.
*  Create a single module containing the glibc _and_ libc++ headers
    * This also seems infeasible, as libc++ already ships with a module map which, again, we can't modify for compatibility reasons.
*  Don't modularize glibc at all; instead, include its headers textually
    * Swift obviously can't include the headers textually; instead, we would build `SwiftGlibc` from a header that wraps the glibc headers.

As I say, we plan to post more details on the forum soon.

Reply:
This is getting addressed in https://github.com/apple/swift/pull/32404.